File: lib/services/home_service.dart
import '../core/constants/app_constants.dart';
import '../core/enums/app_enums.dart';
import '../models/exam/exam_session.dart';
import '../repositories/exam_period_assignment_repository.dart';
import '../repositories/exam_session_repository.dart';

/// ViewModel dashboard — tương đương `DashboardViewModel` trong WebApp.
class DashboardData {
  final String? fullName;
  final String? username;
  final int totalExamSets;
  final int inProgressSessions;
  final int completedSessions;
  final List<ExamSession> recentInProgress;
  final List<ExamSession> recentCompleted;
  final String? errorMessage;

  const DashboardData({
    this.fullName,
    this.username,
    this.totalExamSets = 0,
    this.inProgressSessions = 0,
    this.completedSessions = 0,
    this.recentInProgress = const [],
    this.recentCompleted = const [],
    this.errorMessage,
  });
}

/// Dịch vụ trang chủ — port từ `HomeService.cs`.
class HomeService {
  HomeService(this._assignmentRepo, this._sessionRepo);

  final ExamPeriodAssignmentRepository _assignmentRepo;
  final ExamSessionRepository _sessionRepo;

  Future<DashboardData> buildDashboard({
    required int userId,
    String? fullName,
    String? username,
  }) async {
    var data = DashboardData(fullName: fullName, username: username);
    String? errorMessage;

    final assignmentResult = await _assignmentRepo.getMyAssignments();
    final completedAssignmentIds = <int>{};

    if (assignmentResult.success) {
      final assignments = assignmentResult.data ?? [];
      for (final a in assignments) {
        if (a.status == ExamPeriodAssignmentStatus.completed) {
          completedAssignmentIds.add(a.id);
        }
      }
      data = DashboardData(
        fullName: fullName,
        username: username,
        totalExamSets: assignments.length,
        inProgressSessions: data.inProgressSessions,
        completedSessions: data.completedSessions,
        recentInProgress: data.recentInProgress,
        recentCompleted: data.recentCompleted,
      );
    } else if (assignmentResult.message.isNotEmpty) {
      errorMessage = assignmentResult.message;
    }

    final inProgressResult = await _sessionRepo.getList(
      pageIndex: 1,
      pageSize: 5,
      userId: userId,
      status: ExamSessionStatus.inProgress,
    );

    var recentInProgress = <ExamSession>[];
    if (inProgressResult.success) {
      final raw = inProgressResult.data ?? [];
      recentInProgress = raw
          .where(
            (s) =>
                s.examPeriodAssignmentId == null ||
                !completedAssignmentIds.contains(s.examPeriodAssignmentId),
          )
          .toList();
    }

    final historyResult = await _sessionRepo.getHistory(
      pageIndex: 1,
      userId: userId,
      pageSize: AppConstants.defaultPageSize,
    );

    var recentCompleted = <ExamSession>[];
    var completedCount = 0;
    if (historyResult.success) {
      recentCompleted = historyResult.data ?? [];
      completedCount = historyResult.metaData?.totalItems ??
          recentCompleted.length;
    }

    return DashboardData(
      fullName: fullName,
      username: username,
      totalExamSets: data.totalExamSets,
      inProgressSessions: recentInProgress.length,
      completedSessions: completedCount,
      recentInProgress: recentInProgress,
      recentCompleted: recentCompleted,
      errorMessage: errorMessage,
    );
  }
}

File: ADMIN/Example-API-Flutter/Example.UserService.WebApp/Services/HomeService.cs
using Example.Common.Enums;
using Example.UserService.API.Models;
using Example.UserService.WebApp.Models.Home;
using Example.UserService.WebApp.Repository.IRepository;
using Example.UserService.WebApp.Services.IServices;

namespace Example.UserService.WebApp.Services
{
    /// <summary>
    /// Dịch vụ trang chủ (dashboard): tổng hợp số liệu phân công, phiên đang thi và lịch sử gần đây
    /// từ nhiều Repository, áp dụng quy tắc lọc để ẩn session thuộc phân công đã hoàn thành.
    /// </summary>
    public class HomeService : IHomeService
    {
        private readonly IExamPeriodAssignmentRepository _assignmentRepository;
        private readonly IExamSessionRepository _examSessionRepository;

        // Tiêm repository phân công và phiên thi để tổng hợp số liệu dashboard.
        public HomeService(
            IExamPeriodAssignmentRepository assignmentRepository,
            IExamSessionRepository examSessionRepository)
        {
            _assignmentRepository = assignmentRepository;
            _examSessionRepository = examSessionRepository;
        }

        /// <summary>
        /// Dựng ViewModel dashboard: đếm phân công, lọc session InProgress hợp lệ, và lấy lịch sử gần nhất.
        /// </summary>
        public async Task<DashboardViewModel> BuildDashboardAsync(long userId, string? fullName, string? username)
        {
            var vm = new DashboardViewModel
            {
                FullName = fullName,
                Username = username
            };

            // --- Phân công kỳ thi: tổng số và danh sách Id đã hoàn thành ---
            var assignmentResult = await _assignmentRepository.GetMyAssignmentsAsync();
            var completedAssignmentIds = new HashSet<long>();
            if (assignmentResult.Success)
            {
                vm.TotalExamSets = assignmentResult.Data?.Count ?? 0;
                completedAssignmentIds = (assignmentResult.Data ?? new List<ExamPeriodAssignmentModel>())
                    .Where(x => x.Status == (int)ExamPeriodAssignmentStatusEnum.Completed)
                    .Select(x => x.Id)
                    .ToHashSet();
            }
            else if (string.IsNullOrWhiteSpace(vm.ErrorMessage))
                vm.ErrorMessage = assignmentResult.Message;

            // --- Phiên thi đang làm: tối đa 5 bản ghi, loại session gắn phân công đã Completed ---
            var inProgressResult = await _examSessionRepository.GetListAsync(new ExamSessionSearchModel
            {
                PageIndex = 1,
                PageSize = 5,
                UserId = userId,
                Status = (int)ExamSessionStatusEnum.InProgress
            });
            if (inProgressResult.Success)
            {
                var raw = inProgressResult.Data ?? new List<ExamSessionModel>();
                // Quy tắc nghiệp vụ: không hiển thị bài đang làm nếu phân công tương ứng đã đánh dấu hoàn thành.
                var filtered = raw
                    .Where(s => !s.ExamPeriodAssignmentId.HasValue
                                || !completedAssignmentIds.Contains(s.ExamPeriodAssignmentId.Value))
                    .ToList();

                vm.InProgressSessions = filtered.Count;
                vm.RecentInProgress = filtered;
            }

            // --- Lịch sử thi: 3 phiên gần nhất và tổng số từ metadata phân trang API ---
            var historyResult = await _examSessionRepository.GetHistoryAsync(new ExamSessionSearchModel
            {
                PageIndex = 1,
                PageSize = 3,
                UserId = userId
            });
            if (historyResult.Success)
            {
                vm.CompletedSessions = historyResult.MetaData?.TotalItems ?? historyResult.Data?.Count ?? 0;
                vm.RecentCompleted = historyResult.Data ?? new List<ExamSessionModel>();
            }

            return vm;
        }
    }
}
